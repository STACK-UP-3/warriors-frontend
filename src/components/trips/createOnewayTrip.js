import React from 'react';
import Navbar from '../../components/navBar';
import SideMenu from '../common/sideMenu';
import MobileSidebar from '../common/mobileSidebar';
import Footer from '../footer';
import OnewayTripForm from './onewayTripForm';

export default () => {
  /* istanbul ignore next */
  !window.localStorage.getItem('token')
    ? window.location.assign('/login')
    : null;
  return (
    <div>
      <Navbar />
      <main className='onewayTripContainer'>
        <div className='sideMenu'>
          <div className='fixed-sidebar container center hide-on-med-and-down'>
            <SideMenu />
          </div>
          <div className='show-on-med-and-down container center'>
            <MobileSidebar />
          </div>
        </div>
        <div className='mainContent'>
          <div className='row'>
            <div className='col s12 m2'></div>
            <div className='col s12 m10 content'>
              <h5 className='center'>Create a trip request</h5>
              <OnewayTripForm />
            </div>
          </div>
        </div>
      </main>
      <Footer />
    </div>
  );
};
