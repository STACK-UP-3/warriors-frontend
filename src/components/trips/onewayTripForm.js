import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import SyncLoader from 'react-spinners/SyncLoader';
import { toast, ToastContainer } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import M from 'materialize-css';
import { fetchCities } from '../../redux/actions/citiesAction';
import { onCreateTrip } from '../../redux/actions/tripActions';
import dataValidate from '../../helpers/tripDataValidate';

export class OnewayTripForm extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
  async componentDidMount() {
    await this.props.citiesProcess();
    M.AutoInit();
  }

  onInputChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    const tripData = {
      origin: this.state.origin,
      destination: this.state.destination,
      date: this.state.date,
      returnDate: this.state.returnDate,
      travelReason: this.state.travelReason,
      accommodationID: this.state.accommodationID || 0,
    };
    const validated = dataValidate(tripData);
    if (validated.error) {
      this.setState({
        error: true,
        message: validated.message,
      });
    } else {
      this.setState({
        error: false,
      });

      await this.props.tripProcess(tripData);

      /* istanbul ignore next */
      if (this.props.tripResponse.error) {
      } else {
        document.getElementById('onewayTripForm').reset();
        this.setState({
          origin: null,
          destination: null,
          date: null,
          returnDate: null,
          travelReason: null,
        });
        toast.info(this.props.tripResponse.payload.message);
      }
    }
  };

  render() {
    const cities = this.props.cities.data || [];
    return (
      <div>
        <form
          id='onewayTripForm'
          className='col s12 container center'
          testdata='onewayTripForm'
          onSubmit={this.handleSubmit}>
          <div className='form-content'>
            <div className='row' id='textInput'>
              <div className='input-field col s12'>
                <select
                  name='origin'
                  className='validate'
                  disabled={this.props.tripResponse.loading}
                  onChange={this.onInputChange}>
                  <option defaultValue='none'>Choose your origin city</option>
                  {cities.map((value) => (
                    <option key={value.id} value={value.city}>
                      {value.city}
                    </option>
                  ))}
                </select>
                <label htmlFor='origin'>Your Origin</label>
              </div>
            </div>
            <div className='row' id='textInput'>
              <div
                className='input-field col s12'
                style={{ marginTop: '20px' }}>
                <select
                  name='destination'
                  className='validate'
                  disabled={this.props.tripResponse.loading}
                  onChange={this.onInputChange}>
                  <option defaultValue='none'>
                    Choose your destination city
                  </option>
                  {cities.map((value) => (
                    <option key={value.id} value={value.city}>
                      {value.city}
                    </option>
                  ))}
                </select>
                <label>Your Destination</label>
              </div>
            </div>
            <div className='row' id='textInput'>
              <div
                className='input-field col s12 m6'
                style={{ margin: '20px 0px' }}>
                <input
                  name='date'
                  id='date'
                  type='date'
                  className='validate'
                  testdata='inputChange'
                  onChange={this.onInputChange}
                  disabled={this.props.tripResponse.loading}
                />
                <label htmlFor='date'>Leaving Date</label>
              </div>

              <div
                className='input-field col s12 m6'
                style={{ margin: '20px 0px' }}>
                <input
                  name='returnDate'
                  id='returnDate'
                  type='date'
                  className='validate'
                  testdata='inputChange'
                  onChange={this.onInputChange}
                  disabled={this.props.tripResponse.loading}
                />
                <label htmlFor='returnDate'>Return Date</label>
              </div>
            </div>

            <div className='row' id='textInput'>
              <div className='input-field col s12'>
                <input
                  name='travelReason'
                  id='travelReason'
                  type='text'
                  className='validate'
                  testdata='inputChange'
                  onChange={this.onInputChange}
                  disabled={this.props.tripResponse.loading}
                />
                <label htmlFor='travelReason'>Travel Reason</label>
              </div>
            </div>

            <div className='contentFooter'>
              {this.state.error ? (
                <p className='error'>{this.state.message}</p>
              ) : this.props.tripResponse.error ? (
                <p className='error'>{this.props.tripResponse.message}</p>
              ) : (
                <ToastContainer
                  position='top-center'
                  autoClose={false}
                  hideProgressBar={true}
                  newestOnTop={false}
                  closeOnClick
                  autoClose={8000}
                  rtl={false}
                  pauseOnFocusLoss
                  draggable={false}
                  pauseOnHover
                />
              )}
            </div>
          </div>

          <div className='form-footer'>
            <div className='submitBtn'>
              <button
                className='btn waves-effect waves-light'
                type='submit'
                id='submit'
                disabled={
                  !this.state.origin ||
                  !this.state.destination ||
                  !this.state.date ||
                  !this.state.travelReason ||
                  this.props.tripResponse.loading
                }>
                {this.props.tripResponse.loading ? (
                  <SyncLoader
                    size={5}
                    margin={5}
                    color={'#fff'}
                    loading={this.props.tripResponse.loading}
                  />
                ) : (
                  'Send Request'
                )}
              </button>
            </div>
          </div>
        </form>
      </div>
    );
  }
}

OnewayTripForm.propTypes = {
  tripProcess: PropTypes.func.isRequired,
  citiesProcess: PropTypes.func.isRequired,
};

const mapStateToProps = (response) => ({
  cities: response.cities,
  tripResponse: response.onewayTrip,
});
/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
  citiesProcess: () => dispatch(fetchCities()),
  tripProcess: (tripData) => dispatch(onCreateTrip(tripData)),
});

export default connect(mapStateToProps, mapDispatchToProps)(OnewayTripForm);
