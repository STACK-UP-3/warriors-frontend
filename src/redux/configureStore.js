import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import promise from 'redux-promise-middleware';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducersCombination from './reducers/index';



export default ()=>{
    const store = createStore(
        reducersCombination, {}, 
        composeWithDevTools(applyMiddleware(thunk,promise))
    );
    
    return store;
}
