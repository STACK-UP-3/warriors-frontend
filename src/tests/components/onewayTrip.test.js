import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import mockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import OnewayTripComponent from '../../components/trips/createOnewayTrip';
import OnewayTripFormDefault from '../../components/trips/onewayTripForm';
import { OnewayTripForm } from '../../components/trips/onewayTripForm';
import configureStore from '../../redux/configureStore';

describe('Test <create onway trip/> Component', () => {
  it('Should render component correctly', () => {
    const tripComponent = shallow(<OnewayTripComponent />);
    expect(toJson(tripComponent)).toMatchSnapshot();
    expect(tripComponent.length).toBe(1);
  });
});

describe('Test <oneway trip form/> Component', () => {
  const store = configureStore();
  let onFormSubmit;
  beforeEach(() => {
    onFormSubmit = jest.fn();
  });

  it('Should render the oneway trip form correctly', () => {
    const store = mockStore([thunk])({
      onewayTrip: {
        loading: false,
        error: false,
        message: '',
      },
      cities: [],
    });

    const wrapper = mount(
      <Provider store={store}>
        <OnewayTripFormDefault />
      </Provider>,
    );

    const handleSubmit = jest.spyOn(
      wrapper.find('[testdata="onewayTripForm"]').props(),
      'onSubmit',
    );
    wrapper.find('[testdata="onewayTripForm"]').props().onSubmit();
    expect(wrapper).toMatchSnapshot();
    expect(handleSubmit).toHaveBeenLastCalledWith();
  });

  it('Should test form inputs on change', () => {
    const wrapper = mount(
      <Provider store={store}>
        <OnewayTripFormDefault />
      </Provider>,
    );

    wrapper
      .find('select')
      .at(0)
      .simulate('change', {
        target: { id: 'origin', value: 'bujumbura', name: 'origin' },
      });
    wrapper
      .find('select')
      .at(1)
      .simulate('change', {
        target: { id: 'destination', value: 'kigali', name: 'destination' },
      });
    wrapper
      .find('input')
      .at(0)
      .simulate('change', {
        target: { id: 'date', value: '2020-10-10', name: 'date' },
      });

    wrapper
      .find('input')
      .at(1)
      .simulate('change', {
        target: { id: 'returnDate', value: '2020-10-11', name: 'returnDate' },
      });
    wrapper
      .find('input')
      .at(2)
      .simulate('change', {
        target: { id: 'travelReason', value: 'vacation', name: 'travelReason' },
      });

    expect(wrapper.length).toBe(1);
  });

  it('Should handle form submit process', () => {
    const form = shallow(
      <OnewayTripForm
        citiesProcess={jest.fn()}
        tripProcess={jest.fn()}
        cities={{
          data: [
            { id: 1, city: 'kigali' },
            { id: 1, city: 'bujumbura' },
          ],
        }}
        tripResponse={{ loading: false, error: false }}
      />,
    );
    const env = {
      preventDefault: jest.fn(),
    };
    const instance = form.instance();
    instance.setState({
      origin: 'kigali',
      destination: 'bujumbura',
      date: '2020-10-10',
      error: false,
      returnDate: '2020-10-11',
    });
    instance.handleSubmit(env);
    expect(instance.state).toEqual({
      origin: 'kigali',
      destination: 'bujumbura',
      date: '2020-10-10',
      error: false,
      returnDate: '2020-10-11',
    });
  });
  it('Should handle form submit process', () => {
    const form2 = shallow(
      <OnewayTripForm
        citiesProcess={jest.fn()}
        tripProcess={jest.fn()}
        cities={{
          data: [
            { id: 1, city: 'kigali' },
            { id: 1, city: 'bujumbura' },
          ],
        }}
        tripResponse={{ loading: true, error: true }}
      />,
    );
    const env = {
      preventDefault: jest.fn(),
    };
    const instance = form2.instance();
    instance.setState({
      origin: 'kigali',
      destination: 'bujumbura',
      date: '2020-10-10',
      error: false,
      returnDate: '2020-10-11',
    });
    instance.handleSubmit(env);
    expect(instance.state).toEqual({
      origin: 'kigali',
      destination: 'bujumbura',
      date: '2020-10-10',
      error: false,
      returnDate: '2020-10-11',
    });
  });
  it('Should handle form submit process', () => {
    const form3 = shallow(
      <OnewayTripForm
        citiesProcess={jest.fn()}
        tripProcess={jest.fn()}
        cities={{
          data: [
            { id: 1, city: 'kigali' },
            { id: 1, city: 'bujumbura' },
          ],
        }}
        tripResponse={{ loading: true, error: true }}
      />,
    );
    const env = {
      preventDefault: jest.fn(),
    };
    const instance = form3.instance();
    instance.setState({
      origin: 'kigali',
      destination: 'kigali',
      date: '2020-10-10',
      error: false,
      returnDate: '2020-10-11',
    });
    instance.handleSubmit(env);
    expect(instance.state.error).toEqual(true);
  });
  it('Should handle form submit process', () => {
    const form4 = shallow(
      <OnewayTripForm
        citiesProcess={jest.fn()}
        tripProcess={jest.fn()}
        cities={{
          data: [
            { id: 1, city: 'kigali' },
            { id: 1, city: 'bujumbura' },
          ],
        }}
        tripResponse={{ loading: true, error: true }}
      />,
    );
    const env = {
      preventDefault: jest.fn(),
    };
    const instance = form4.instance();
    instance.setState({
      origin: 'kigali',
      destination: 'bujumbura',
      date: '2020-10-10',
      error: false,
      returnDate: '2020-09-09',
    });
    instance.handleSubmit(env);
    expect(instance.state.error).toEqual(true);
  });
  it('Should handle form submit process', () => {
    const form5 = shallow(
      <OnewayTripForm
        citiesProcess={jest.fn()}
        tripProcess={jest.fn()}
        cities={{
          data: [
            { id: 1, city: 'kigali' },
            { id: 1, city: 'bujumbura' },
          ],
        }}
        tripResponse={{ loading: true, error: true }}
      />,
    );
    const env = {
      preventDefault: jest.fn(),
    };
    const instance = form5.instance();
    instance.setState({
      origin: 'kigali',
      destination: 'bujumbura',
      date: '2020-02-02',
      error: false,
      returnDate: '2020-09-09',
    });
    instance.handleSubmit(env);
    expect(instance.state.error).toEqual(true);
  });
});
