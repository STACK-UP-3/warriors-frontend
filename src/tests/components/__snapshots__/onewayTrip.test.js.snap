// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <create onway trip/> Component Should render component correctly 1`] = `
<div>
  <NavBar />
  <main
    className="onewayTripContainer"
  >
    <div
      className="sideMenu"
    >
      <div
        className="fixed-sidebar container center hide-on-med-and-down"
      >
        <_default />
      </div>
      <div
        className="show-on-med-and-down container center"
      >
        <mobileSidebar />
      </div>
    </div>
    <div
      className="mainContent"
    >
      <div
        className="row"
      >
        <div
          className="col s12 m2"
        />
        <div
          className="col s12 m10 content"
        >
          <h5
            className="center"
          >
            Create a trip request
          </h5>
          <Connect(OnewayTripForm) />
        </div>
      </div>
    </div>
  </main>
  <_default />
</div>
`;

exports[`Test <oneway trip form/> Component Should render the oneway trip form correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(OnewayTripForm)>
    <OnewayTripForm
      cities={Array []}
      citiesProcess={[Function]}
      tripProcess={[Function]}
      tripResponse={
        Object {
          "error": false,
          "loading": false,
          "message": "",
        }
      }
    >
      <div>
        <form
          className="col s12 container center"
          id="onewayTripForm"
          onSubmit={
            [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            }
          }
          testdata="onewayTripForm"
        >
          <div
            className="form-content"
          >
            <div
              className="row"
              id="textInput"
            >
              <div
                className="input-field col s12"
              >
                <select
                  className="validate"
                  disabled={false}
                  name="origin"
                  onChange={[Function]}
                >
                  <option
                    defaultValue="none"
                  >
                    Choose your origin city
                  </option>
                </select>
                <label
                  htmlFor="origin"
                >
                  Your Origin
                </label>
              </div>
            </div>
            <div
              className="row"
              id="textInput"
            >
              <div
                className="input-field col s12"
                style={
                  Object {
                    "marginTop": "20px",
                  }
                }
              >
                <select
                  className="validate"
                  disabled={false}
                  name="destination"
                  onChange={[Function]}
                >
                  <option
                    defaultValue="none"
                  >
                    Choose your destination city
                  </option>
                </select>
                <label>
                  Your Destination
                </label>
              </div>
            </div>
            <div
              className="row"
              id="textInput"
            >
              <div
                className="input-field col s12 m6"
                style={
                  Object {
                    "margin": "20px 0px",
                  }
                }
              >
                <input
                  className="validate"
                  disabled={false}
                  id="date"
                  name="date"
                  onChange={[Function]}
                  testdata="inputChange"
                  type="date"
                />
                <label
                  htmlFor="date"
                >
                  Leaving Date
                </label>
              </div>
              <div
                className="input-field col s12 m6"
                style={
                  Object {
                    "margin": "20px 0px",
                  }
                }
              >
                <input
                  className="validate"
                  disabled={false}
                  id="returnDate"
                  name="returnDate"
                  onChange={[Function]}
                  testdata="inputChange"
                  type="date"
                />
                <label
                  htmlFor="returnDate"
                >
                  Return Date
                </label>
              </div>
            </div>
            <div
              className="row"
              id="textInput"
            >
              <div
                className="input-field col s12"
              >
                <input
                  className="validate"
                  disabled={false}
                  id="travelReason"
                  name="travelReason"
                  onChange={[Function]}
                  testdata="inputChange"
                  type="text"
                />
                <label
                  htmlFor="travelReason"
                >
                  Travel Reason
                </label>
              </div>
            </div>
            <div
              className="contentFooter"
            >
              <ToastContainer
                autoClose={8000}
                closeButton={[Function]}
                closeOnClick={true}
                draggable={false}
                draggablePercent={80}
                hideProgressBar={true}
                newestOnTop={false}
                pauseOnFocusLoss={true}
                pauseOnHover={true}
                position="top-center"
                role="alert"
                rtl={false}
                transition={[Function]}
              >
                <div
                  className="Toastify"
                />
              </ToastContainer>
            </div>
          </div>
          <div
            className="form-footer"
          >
            <div
              className="submitBtn"
            >
              <button
                className="btn waves-effect waves-light"
                disabled={true}
                id="submit"
                type="submit"
              >
                Send Request
              </button>
            </div>
          </div>
        </form>
      </div>
    </OnewayTripForm>
  </Connect(OnewayTripForm)>
</Provider>
`;
